### PR TITLE
use matches!() macro for simple if let conditions

### PR DESCRIPTION
**Pull request by matthiaskrgr**

### CODE REVIEW

File: compiler/rustc_ast/src/ast.rs [hunk 1]
```diff
@@ -1931,7 +1931,7 @@ pub enum TyKind {
 
 impl TyKind {
     pub fn is_implicit_self(&self) -> bool {
-        if let TyKind::ImplicitSelf = *self { true } else { false }
+        matches!(*self, TyKind::ImplicitSelf)
```
Comment (lcnr) [inline compiler/rustc_ast/src/ast.rs:1934]:
```suggestion
        matches!(self, TyKind::ImplicitSelf)
```

We should also change `TyKind` to be `Copy`... will just throw a PR up for that :shrug:

File: compiler/rustc_ast/src/ast.rs [hunk 2]
```diff
@@ -2508,7 +2508,7 @@ pub enum VisibilityKind {
 
 impl VisibilityKind {
     pub fn is_pub(&self) -> bool {
-        if let VisibilityKind::Public = *self { true } else { false }
+        matches!(*self, VisibilityKind::Public)
```
Comment (lcnr) [inline compiler/rustc_ast/src/ast.rs:2511]:
```suggestion
        matches!(self, VisibilityKind::Public)
```

File: compiler/rustc_ast_passes/src/ast_validation.rs [hunk 3]
```diff
@@ -868,10 +868,7 @@ impl<'a> Visitor<'a> for AstValidator<'a> {
                     .emit();
                 }
 
-                if !bounds
-                    .iter()
-                    .any(|b| if let GenericBound::Trait(..) = *b { true } else { false })
-                {
+                if !bounds.iter().any(|b| matches!(*b, GenericBound::Trait(..))) {
```
Comment (lcnr) [inline compiler/rustc_ast_passes/src/ast_validation.rs:871]:
```suggestion
                if !bounds.iter().any(|b| matches!(b, GenericBound::Trait(..))) {
```

File: compiler/rustc_attr/src/builtin.rs [hunk 4]
```diff
@@ -160,10 +160,10 @@ pub enum StabilityLevel {
 
 impl StabilityLevel {
     pub fn is_unstable(&self) -> bool {
-        if let StabilityLevel::Unstable { .. } = *self { true } else { false }
+        matches!(*self, StabilityLevel::Unstable { .. })
```
Comment (lcnr) [inline compiler/rustc_attr/src/builtin.rs:163]:
```suggestion
        matches!(self, StabilityLevel::Unstable { .. })
```

File: compiler/rustc_attr/src/builtin.rs [hunk 5]
```diff
@@ -160,10 +160,10 @@ pub enum StabilityLevel {
...
     }
     pub fn is_stable(&self) -> bool {
-        if let StabilityLevel::Stable { .. } = *self { true } else { false }
+        matches!(*self, StabilityLevel::Stable { .. })
```
Comment (lcnr) [inline compiler/rustc_attr/src/builtin.rs:166]:
```suggestion
        matches!(self, StabilityLevel::Stable { .. })
```

File: compiler/rustc_middle/src/middle/cstore.rs [hunk 6]
```diff
@@ -69,7 +69,7 @@ pub enum LibSource {
 
 impl LibSource {
     pub fn is_some(&self) -> bool {
-        if let LibSource::Some(_) = *self { true } else { false }
+        matches!(*self, LibSource::Some(_))
```
Comment (lcnr) [inline compiler/rustc_middle/src/middle/cstore.rs:72]:
```suggestion
        matches!(self, LibSource::Some(_))
```

File: compiler/rustc_mir/src/transform/promote_consts.rs [hunk 7]
```diff
@@ -92,7 +92,7 @@ pub enum TempState {
 impl TempState {
     pub fn is_promotable(&self) -> bool {
         debug!("is_promotable: self={:?}", self);
-        if let TempState::Defined { .. } = *self { true } else { false }
+        matches!(*self, TempState::Defined { .. } )
```
Comment (lcnr) [inline compiler/rustc_mir/src/transform/promote_consts.rs:95]:
```suggestion
        matches!(self, TempState::Defined { .. } )
```

File: compiler/rustc_mir_build/src/build/block.rs [hunk 8]
```diff
@@ -96,8 +96,7 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
                     );
                 }
                 StmtKind::Let { remainder_scope, init_scope, pattern, initializer, lint_level } => {
-                    let ignores_expr_result =
-                        if let PatKind::Wild = *pattern.kind { true } else { false };
+                    let ignores_expr_result = matches!(*pattern.kind, PatKind::Wild);
```
Comment (lcnr) [inline compiler/rustc_mir_build/src/build/block.rs:99]:
```suggestion
                    let ignores_expr_result = matches!(pattern.kind, PatKind::Wild);
```

**Review by lcnr — APPROVED** [review]
LGTM

The same nit 8 times which is mostly personal style. Otherwise r=me I guess

File: compiler/rustc_mir_build/src/build/block.rs (hunk 8 above)
Comment (matthiaskrgr) [reply compiler/rustc_mir_build/src/build/block.rs]: This one is actually needed

````
error[E0308]: mismatched types
   --> compiler/rustc_mir_build/src/build/block.rs:99:70
    |
99  |                     let ignores_expr_result = matches!(pattern.kind, PatKind::Wild);
    |                                                        ------------  ^^^^^^^^^^^^^ expected struct `std::boxed::Box`, found enum `thir::pattern::PatKind`
    |                                                        |
    |                                                        this expression has type `std::boxed::Box<thir::pattern::PatKind<'_>>`
    |
   ::: compiler/rustc_mir_build/src/thir/pattern/mod.rs:121:5
    |
121 |     Wild,
    |     ---- unit variant defined here
    |
    = note: expected struct `std::boxed::Box<thir::pattern::PatKind<'_>>`
                 found enum `thir::pattern::PatKind<'_>`

```

**Comment by lcnr** [comment]
Thanks

@bors r+ rollup

File: compiler/rustc_mir_build/src/build/block.rs (hunk 8 above)
Comment (lcnr) [reply compiler/rustc_mir_build/src/build/block.rs:99]: :+1:

**Pull request merged**